File: examples/lwjgl-map/src/main/java/org/maplibre/nativeffi/examples/lwjglmap/Viewport.java
package org.maplibre.nativeffi.examples.lwjglmap;

import static org.lwjgl.glfw.GLFW.glfwGetFramebufferSize;
import static org.lwjgl.glfw.GLFW.glfwGetWindowContentScale;
import static org.lwjgl.glfw.GLFW.glfwGetWindowSize;

import org.lwjgl.system.MemoryStack;

record Viewport(
    int width,
    int height,
    double scaleFactor,
    int framebufferWidth,
    int framebufferHeight,
    boolean empty) {
  static Viewport read(long window) {
    try (var stack = MemoryStack.stackPush()) {
      var windowWidth = stack.mallocInt(1);
      var windowHeight = stack.mallocInt(1);
      var framebufferWidth = stack.mallocInt(1);
      var framebufferHeight = stack.mallocInt(1);
      var xScale = stack.mallocFloat(1);
      var yScale = stack.mallocFloat(1);
      glfwGetWindowSize(window, windowWidth, windowHeight);
      glfwGetFramebufferSize(window, framebufferWidth, framebufferHeight);
      glfwGetWindowContentScale(window, xScale, yScale);

      var rawLogicalWidth = windowWidth.get(0);
      var rawLogicalHeight = windowHeight.get(0);
      var rawPhysicalWidth = framebufferWidth.get(0);
      var rawPhysicalHeight = framebufferHeight.get(0);
      var empty =
          rawLogicalWidth <= 0
              || rawLogicalHeight <= 0
              || rawPhysicalWidth <= 0
              || rawPhysicalHeight <= 0;

      var logicalWidth = Math.max(1, rawLogicalWidth);
      var logicalHeight = Math.max(1, rawLogicalHeight);
      var physicalWidth = Math.max(1, rawPhysicalWidth);
      var physicalHeight = Math.max(1, rawPhysicalHeight);
      var scale =
          Math.max((double) physicalWidth / logicalWidth, (double) physicalHeight / logicalHeight);
      if (!(scale > 0.0) || !Double.isFinite(scale)) {
        scale = Math.max(xScale.get(0), yScale.get(0));
      }
      if (!(scale > 0.0) || !Double.isFinite(scale)) {
        scale = 1.0;
      }
      return new Viewport(logicalWidth, logicalHeight, scale, physicalWidth, physicalHeight, empty);
    }
  }

  void log(String label) {
    System.out.printf(
        "%s: logical=%dx%d physical=%dx%d scale=%.2f%s%n",
        label,
        width,
        height,
        framebufferWidth,
        framebufferHeight,
        scaleFactor,
        empty ? " empty=true" : "");
  }
}

File: examples/lwjgl-map/src/main/kotlin/org/maplibre/nativeffi/examples/lwjglmap/Viewport.kt
package org.maplibre.nativeffi.examples.lwjglmap

import kotlin.math.ceil
import kotlin.math.max
import org.lwjgl.glfw.GLFW.glfwGetFramebufferSize
import org.lwjgl.glfw.GLFW.glfwGetWindowContentScale
import org.lwjgl.glfw.GLFW.glfwGetWindowSize
import org.lwjgl.system.MemoryStack

internal class Viewport(
  private val width: Int,
  private val height: Int,
  private val scaleFactor: Double,
  private val framebufferWidth: Int,
  private val framebufferHeight: Int,
  private val empty: Boolean,
) {
  fun width(): Int = width

  fun height(): Int = height

  fun scaleFactor(): Double = scaleFactor

  fun framebufferWidth(): Int = framebufferWidth

  fun framebufferHeight(): Int = framebufferHeight

  fun empty(): Boolean = empty

  fun log(label: String) {
    System.out.printf(
      "%s: logical=%dx%d physical=%dx%d scale=%.2f%s%n",
      label,
      width,
      height,
      framebufferWidth,
      framebufferHeight,
      scaleFactor,
      if (empty) " empty=true" else "",
    )
  }

  companion object {
    @JvmStatic
    fun read(window: Long): Viewport {
      MemoryStack.stackPush().use { stack ->
        val windowWidth = stack.mallocInt(1)
        val windowHeight = stack.mallocInt(1)
        val framebufferWidth = stack.mallocInt(1)
        val framebufferHeight = stack.mallocInt(1)
        val xScale = stack.mallocFloat(1)
        val yScale = stack.mallocFloat(1)
        glfwGetWindowSize(window, windowWidth, windowHeight)
        glfwGetFramebufferSize(window, framebufferWidth, framebufferHeight)
        glfwGetWindowContentScale(window, xScale, yScale)

        val rawLogicalWidth = windowWidth[0]
        val rawLogicalHeight = windowHeight[0]
        val rawPhysicalWidth = framebufferWidth[0]
        val rawPhysicalHeight = framebufferHeight[0]
        val empty =
          rawLogicalWidth <= 0 ||
            rawLogicalHeight <= 0 ||
            rawPhysicalWidth <= 0 ||
            rawPhysicalHeight <= 0

        val physicalWidth = max(1, rawPhysicalWidth)
        val physicalHeight = max(1, rawPhysicalHeight)
        var scale = max(xScale[0].toDouble(), yScale[0].toDouble())
        if (!(scale > 0.0) || !scale.isFinite()) {
          scale =
            max(
              physicalWidth.toDouble() / max(1, rawLogicalWidth),
              physicalHeight.toDouble() / max(1, rawLogicalHeight),
            )
        }
        if (!(scale > 0.0) || !scale.isFinite()) {
          scale = 1.0
        }
        val logicalWidth = scaledLogicalSize(physicalWidth, scale)
        val logicalHeight = scaledLogicalSize(physicalHeight, scale)
        return Viewport(logicalWidth, logicalHeight, scale, physicalWidth, physicalHeight, empty)
      }
    }

    private fun scaledLogicalSize(physicalSize: Int, scaleFactor: Double): Int =
      max(1, ceil(physicalSize / scaleFactor).toInt())
  }
}
